| Case ID | Rater A | Rater B |
| :--- | :--- | :--- |
| **01** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-2 |
| **02** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-2, D4-1 |
| **03** | D1-2, D2-1, D3-2, D4-2 | D1-2, D2-1, D3-2, D4-2 |
| **04** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **05** | D1-2, D2-1, D3-2, D4-2 | D1-2, D2-1, D3-2, D4-2 |
| **06** | D1-2, D2-1, D3-1, D4-2 | D1-2, D2-1, D3-2, D4-2 |
| **07** | D1-2, D2-1, D3-1, D4-2 | D1-2, D2-1, D3-1, D4-2 |
| **08** | D1-2, D2-1, D3-1, D4-2 | D1-2, D2-1, D3-1, D4-2 |
| **09** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **10** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **11** | D1-2, D2-2, D3-2, D4-2 | D1-2, D2-1, D3-2, D4-2 |
| **12** | D1-2, D2-1, D3-2, D4-2 | D1-2, D2-1, D3-2, D4-2 |
| **13** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-2, D4-1 |
| **14** | D1-2, D2-1, D3-2, D4-1 | D1-1, D2-1, D3-2, D4-1 |
| **15** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **16** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **17** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **18** | D1-2, D2-1, D3-1, D4-2 | D1-2, D2-1, D3-1, D4-2 |
| **19** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **20** | D1-2, D2-1, D3-1, D4-2 | D1-2, D2-1, D3-1, D4-2 |
| **21** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **22** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **23** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-2, D4-1 |
| **24** | D1-2, D2-2, D3-2, D4-2 | D1-2, D2-2, D3-2, D4-2 |
| **25** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-2, D4-1 |
| **26** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **27** | D1-2, D2-1, D3-1, D4-2 | D1-2, D2-1, D3-1, D4-2 |
| **28** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **29** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **30** | D1-2, D2-1, D3-1, D4-1 | D1-1, D2-1, D3-1, D4-1 |
| **31** | D1-2, D2-1, D3-2, D4-1 | D1-2, D2-1, D3-1, D4-1 |
| **32** | D1-2, D2-1, D3-1, D4-1 | D1-2, D2-1, D3-1, D4-1 |

In [1]:
import pandas as pd
import numpy as np
import warnings
from sklearn.metrics import cohen_kappa_score

# -------------------------------------------------------------------
# 1. DATASET (Pre-Discussion Independent Scores)
# -------------------------------------------------------------------
rater_a_raw = {
    "01": [2, 1, 1, 1], "02": [2, 1, 1, 1], "03": [2, 1, 2, 2], "04": [2, 1, 1, 1],
    "05": [2, 1, 2, 2], "06": [2, 1, 1, 2], "07": [2, 1, 1, 2], "08": [2, 1, 1, 2],
    "09": [2, 1, 1, 1], "10": [2, 1, 1, 1], "11": [2, 2, 2, 2], "12": [2, 1, 2, 2],
    "13": [2, 1, 1, 1], "14": [2, 1, 2, 1], "15": [2, 1, 1, 1], "16": [2, 1, 1, 1],
    "17": [2, 1, 1, 1], "18": [2, 1, 1, 2], "19": [2, 1, 1, 1], "20": [2, 1, 1, 2],
    "21": [2, 1, 1, 1], "22": [2, 1, 1, 1], "23": [2, 1, 1, 1], "24": [2, 2, 2, 2],
    "25": [2, 1, 1, 1], "26": [2, 1, 1, 1], "27": [2, 1, 1, 2], "28": [2, 1, 1, 1],
    "29": [2, 1, 1, 1], "30": [2, 1, 1, 1], "31": [2, 1, 2, 1], "32": [2, 1, 1, 1]
}

rater_b_raw = {
    "01": [2, 1, 1, 2], "02": [2, 1, 2, 1], "03": [2, 1, 2, 2], "04": [2, 1, 1, 1],
    "05": [2, 1, 2, 2], "06": [2, 1, 2, 2], "07": [2, 1, 1, 2], "08": [2, 1, 1, 2],
    "09": [2, 1, 1, 1], "10": [2, 1, 1, 1], "11": [2, 1, 2, 2], "12": [2, 1, 2, 2],
    "13": [2, 1, 2, 1], "14": [1, 1, 2, 1], "15": [2, 1, 1, 1], "16": [2, 1, 1, 1],
    "17": [2, 1, 1, 1], "18": [2, 1, 1, 2], "19": [2, 1, 1, 1], "20": [2, 1, 1, 2],
    "21": [2, 1, 1, 1], "22": [2, 1, 1, 1], "23": [2, 1, 2, 1], "24": [2, 2, 2, 2],
    "25": [2, 1, 2, 1], "26": [2, 1, 1, 1], "27": [2, 1, 1, 2], "28": [2, 1, 1, 1],
    "29": [2, 1, 1, 1], "30": [1, 1, 1, 1], "31": [2, 1, 1, 1], "32": [2, 1, 1, 1]
}

dimensions = ["D1", "D2", "D3", "D4"]
cases = sorted(rater_a_raw.keys())

# -------------------------------------------------------------------
# 2. PER-DIMENSION CALCULATIONS
# -------------------------------------------------------------------
print("="*75)
print("COHEN'S KAPPA -- PER DIMENSION".center(75))
print("="*75)

for i, dim in enumerate(dimensions):
    ra_dim = [rater_a_raw[c][i] for c in cases]
    rb_dim = [rater_b_raw[c][i] for c in cases]

    exact_matches = sum(1 for a, b in zip(ra_dim, rb_dim) if a == b)
    total = len(cases)

    # If all values are identical, variance=0 will occur; suppress unnecessary warnings
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        kappa = cohen_kappa_score(ra_dim, rb_dim, labels=[1, 2])

    if np.isnan(kappa) or (exact_matches == total and len(set(ra_dim)) == 1):
        kappa_str = "Undefined (perfect agreement, variance = 0)"
    else:
        kappa_str = f"{kappa:.4f}"

    print(f"{dim}:  kappa = {kappa_str:<45} exact agreement = {exact_matches}/{total}")

# -------------------------------------------------------------------
# 3. OVERALL POOLED CALCULATION
# -------------------------------------------------------------------
flat_rater_a = [val for c in cases for val in rater_a_raw[c]]
flat_rater_b = [val for c in cases for val in rater_b_raw[c]]

overall_matches = sum(1 for a, b in zip(flat_rater_a, flat_rater_b) if a == b)
overall_total = len(flat_rater_a)
overall_kappa = cohen_kappa_score(flat_rater_a, flat_rater_b)

print("\n" + "="*75)
print("OVERALL KAPPA (ALL DIMENSIONS POOLED)".center(75))
print("="*75)
print(f"Overall:  kappa = {overall_kappa:.4f}   exact agreement = {overall_matches}/{overall_total}")

# -------------------------------------------------------------------
# 4. DISAGREEMENT MATRIX (CASE LEVEL TABLE)
# -------------------------------------------------------------------
print("\n" + "="*75)
print("DISAGREEMENT TABLE (CASE LEVEL)".center(75))
print("="*75)
print(" Case | R1 R2 | R1 R2 | R1 R2 | R1 R2 | Status")
print("      |  D1   |  D2   |  D3   |  D4   |")
print("-" * 75)

for case_id in cases:
    scores_a = rater_a_raw[case_id]
    scores_b = rater_b_raw[case_id]

    col_strs = []
    disagreements = []

    for i in range(4):
        val_a = scores_a[i]
        val_b = scores_b[i]
        col_strs.append(f" {val_a}  {val_b} ")
        if val_a != val_b:
            disagreements.append(dimensions[i])

    cols_joined = "|".join(col_strs)

    if not disagreements:
        status = "Full agreement"
    else:
        status = "Disagreement: " + ", ".join(disagreements)

    print(f"  {case_id}  |{cols_joined}| {status}")

# -------------------------------------------------------------------
# 5. ACADEMIC REFERENCE AND AUTOMATIC INTERPRETATION
# -------------------------------------------------------------------
def interpret_kappa(k):
    if k > 0.80: return "Almost Perfect"
    elif k > 0.60: return "Substantial"
    elif k > 0.40: return "Moderate"
    elif k > 0.20: return "Fair"
    elif k > 0.00: return "Slight"
    else: return "Poor"

print("\n" + "="*75)
print(f"CONCLUSION: Overall agreement is '{interpret_kappa(overall_kappa)}' ({overall_kappa:.4f})")
print("="*75)

print("\nKappa interpretation guide (Landis & Koch, 1977):")
print("  < 0.00        : Poor")
print("  0.00 - 0.20   : Slight")
print("  0.21 - 0.40   : Fair")
print("  0.41 - 0.60   : Moderate")
print("  0.61 - 0.80   : Substantial")
print("  0.81 - 1.00   : Almost Perfect")

                       COHEN'S KAPPA -- PER DIMENSION                      
D1:  kappa = 0.0000                                        exact agreement = 30/32
D2:  kappa = 0.6522                                        exact agreement = 31/32
D3:  kappa = 0.5450                                        exact agreement = 26/32
D4:  kappa = 0.9322                                        exact agreement = 31/32

                   OVERALL KAPPA (ALL DIMENSIONS POOLED)                   
Overall:  kappa = 0.8390   exact agreement = 118/128

                      DISAGREEMENT TABLE (CASE LEVEL)                      
 Case | R1 R2 | R1 R2 | R1 R2 | R1 R2 | Status
      |  D1   |  D2   |  D3   |  D4   |
---------------------------------------------------------------------------
  01  | 2  2 | 1  1 | 1  1 | 1  2 | Disagreement: D4
  02  | 2  2 | 1  1 | 1  2 | 1  1 | Disagreement: D3
  03  | 2  2 | 1  1 | 2  2 | 2  2 | Full agreement
  04  | 2  2 | 1  1 | 1  1 | 1  1 | Full agreement
  05  | 2  2 |

## Inter-Rater Reliability Analysis and Consensus Resolution

Two independent raters initially scored 32 evaluation units across four dimensions (D1, D2, D3, D4) using an ordinal rating scale. Inter-rater reliability was calculated using **Cohen's Kappa** based on these pre-discussion independent scores.

The overall initial performance across all dimensions pooled yielded a kappa of **0.8390** (118/128 exact agreement), which falls within the *Almost Perfect* range according to the Landis & Koch (1977) classification.

When analyzed per dimension, **D4** demonstrated near-perfect reliability with a kappa of **0.93** (31/32 exact agreement). **D2** yielded a *Substantial* kappa of **0.65** (31/32 exact agreement), while **D3** showed *Moderate* agreement with a kappa of **0.54** (26/32 exact agreement).

**Methodological Note on D1:**
For **D1**, the exact agreement between raters was notably high (30 out of 32 cases). However, the calculated statistical kappa value is **0.00**. This is a recognized statistical phenomenon known as the *Kappa prevalence/base-rate paradox* (Feinstein & Cicchetti, 1990). Because almost all cases fell into a single dominant scoring category (a score of 2), the variance was extremely low. In highly skewed marginal distributions like this, the kappa statistic artificially collapses to zero despite highly consistent actual agreement. Therefore, the exact agreement metric (93.75%) is utilized as the primary reflection of inter-rater reliability for this specific dimension instead of the statistical kappa.

| Dimension | Pre-Discussion Kappa | Exact Agreement | Classification |
| :--- | :---: | :---: | :--- |
| Dimension 1 (D1) | 0.00* | 30/32 | *Paradoxical zero (High agreement, low variance)* |
| Dimension 2 (D2) | 0.65 | 31/32 | Substantial |
| Dimension 3 (D3) | 0.54 | 26/32 | Moderate |
| Dimension 4 (D4) | 0.93 | 31/32 | Almost Perfect |
| **Overall (pooled)** | **0.8390** | **118/128** | **Almost Perfect** |

*Reference: Landis & Koch (1977). Biometrics, 33(1), 159–174.*

---

### Disagreement Pattern and Final Consensus Calibration

During the initial independent scoring phase, 22 of the 32 evaluation units reached full agreement across all four dimensions. The remaining 10 cases showed a discrepancy strictly in exactly one dimension each; no single case had simultaneous disagreements across multiple dimensions. Furthermore, all documented disagreements occurred strictly between adjacent categories (1 vs 2), with no extreme discrepancies observed. This indicates that the raters share a fundamentally aligned understanding of the evaluation rubric, with deviations stemming only from minor threshold interpretations.

The pre-discussion discrepancies were distributed as follows:
* **D3** (6 cases — Cases 02, 06, 13, 23, 25, 31): Accounted for the majority of the variance. In 5 out of these 6 cases, Rater 2 demonstrated a systematic tendency to score higher (2) compared to Rater 1 (1).
* **D1** (2 cases — Cases 14, 30): Rater 1 = 2, Rater 2 = 1.
* **D2** (1 case — Case 11): Rater 1 = 2, Rater 2 = 1.
* **D4** (1 case — Case 01): Rater 1 = 1, Rater 2 = 2.

**Final Resolution:**
Following the independent analyses, the raters held a consensus session to resolve these 10 single-dimension discrepancies. Recognizing that 60% of the variations occurred specifically in **D3**, a targeted rubric calibration was performed during this session to realign the scoring thresholds for that dimension.

Through this collaborative and methodological review, final consensus scores were established for all discrepant items. The dataset utilized for the analytical runs reflects this 100% post-discussion consensus.
Of the 32 evaluation units, 22 reached the final consensus table without requiring discussion, as both raters had assigned identical scores independently (Cases 03, 04, 05, 07, 08, 09, 10, 12, 15, 16, 17, 18, 19, 20, 21, 22, 24, 26, 27, 28, 29, 32). The remaining 10 cases were resolved through the consensus session. In 6 of these, the final score aligned with Rater A's initial judgment (Cases 02, 06, 13, 14, 23, 30), and in 4 cases with Rater B's (Cases 01, 11, 25, 31). Notably, Cases 02, 13, and 23 — which correspond to Gemma 4 Sanity Run evaluations — were among those resolved in favor of Rater A.
The final score table is presented below:

---

### Final Consensus Dataset

| Case | D1 | D2 | D3 | D4 | Total | Prompt ID | Model | Run Type | Internal Logic |
| :--- | :---: | :---: | :---: | :---: | :---: | :--- | :--- | :--- | :--- |
| **01** | 2 | 1 | 1 | 2 | 6/8 | P03 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **02** | 2 | 1 | 1 | 1 | 5/8 | P01 | Gemma 4 | Sanity Run 1 | Gemma 4 - First repeat EN-TR pair |
| **03** | 2 | 1 | 2 | 2 | 7/8 | P05 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **04** | 2 | 1 | 1 | 1 | 5/8 | P08 | Gemma 4 | Sanity Run 2 | Gemma 4 - Second repeat EN-TR pair |
| **05** | 2 | 1 | 2 | 2 | 7/8 | P02 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **06** | 2 | 1 | 1 | 2 | 6/8 | P04 | ChatGPT | Sanity Run 1 | ChatGPT - First repeat EN-TR pair |
| **07** | 2 | 1 | 1 | 2 | 6/8 | P01 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **08** | 2 | 1 | 1 | 2 | 6/8 | P06 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **09** | 2 | 1 | 1 | 1 | 5/8 | P05 | Gemma 4 | Sanity Run 2 | Gemma 4 - Second repeat EN-TR pair |
| **10** | 2 | 1 | 1 | 1 | 5/8 | P04 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **11** | 2 | 1 | 2 | 2 | 7/8 | P08 | ChatGPT | Sanity Run 1 | ChatGPT - First repeat EN-TR pair |
| **12** | 2 | 1 | 2 | 2 | 7/8 | P07 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **13** | 2 | 1 | 1 | 1 | 5/8 | P01 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **14** | 2 | 1 | 2 | 1 | 6/8 | P05 | ChatGPT | Sanity Run 1 | ChatGPT - First repeat EN-TR pair |
| **15** | 2 | 1 | 1 | 1 | 5/8 | P03 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **16** | 2 | 1 | 1 | 1 | 5/8 | P04 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **17** | 2 | 1 | 1 | 1 | 5/8 | P08 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **18** | 2 | 1 | 1 | 2 | 6/8 | P01 | ChatGPT | Sanity Run 2 | ChatGPT - Second repeat EN-TR pair |
| **19** | 2 | 1 | 1 | 1 | 5/8 | P05 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **20** | 2 | 1 | 1 | 2 | 6/8 | P02 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **21** | 2 | 1 | 1 | 1 | 5/8 | P04 | Gemma 4 | Sanity Run 2 | Gemma 4 - Second repeat EN-TR pair |
| **22** | 2 | 1 | 1 | 1 | 5/8 | P08 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **23** | 2 | 1 | 1 | 1 | 5/8 | P01 | Gemma 4 | Sanity Run 2 | Gemma 4 - Second repeat EN-TR pair |
| **24** | 2 | 2 | 2 | 2 | 8/8 | P05 | ChatGPT | Sanity Run 2 | ChatGPT - Second repeat EN-TR pair |
| **25** | 2 | 1 | 2 | 1 | 6/8 | P07 | Gemma 4 | Original | Gemma 4 - Original EN-TR pair |
| **26** | 2 | 1 | 1 | 1 | 5/8 | P04 | Gemma 4 | Sanity Run 1 | Gemma 4 - First repeat EN-TR pair |
| **27** | 2 | 1 | 1 | 2 | 6/8 | P06 | ChatGPT | Original | ChatGPT - Original EN-TR pair |
| **28** | 2 | 1 | 1 | 1 | 5/8 | P08 | Gemma 4 | Sanity Run 1 | Gemma 4 - First repeat EN-TR pair |
| **29** | 2 | 1 | 1 | 1 | 5/8 | P01 | ChatGPT | Sanity Run 1 | ChatGPT - First repeat EN-TR pair |
| **30** | 2 | 1 | 1 | 1 | 5/8 | P04 | ChatGPT | Sanity Run 2 | ChatGPT - Second repeat EN-TR pair |
| **31** | 2 | 1 | 1 | 1 | 5/8 | P08 | ChatGPT | Sanity Run 2 | ChatGPT - Second repeat EN-TR pair |
| **32** | 2 | 1 | 1 | 1 | 5/8 | P05 | Gemma 4 | Sanity Run 1 | Gemma 4 - First repeat EN-TR pair |